python/ql/src/experimental/semmle/python/frameworks/ClickHouseDriver.qll

```diff
@@ -0,0 +1,85 @@
+/**
+ * Provides classes modeling security-relevant aspects of `clickhouse-driver` and `aioch` PyPI packages.
+ * See
+ * - https://pypi.org/project/clickhouse-driver/
+ * - https://pypi.org/project/aioch/
+ * - https://clickhouse-driver.readthedocs.io/en/latest/
+ */
+
+private import python
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+private import semmle.python.frameworks.PEP249
+
+/**
+ * Provides models for `clickhouse-driver` and `aioch` PyPI packages.
+ * See
+ * - https://pypi.org/project/clickhouse-driver/
+ * - https://pypi.org/project/aioch/
+ * - https://clickhouse-driver.readthedocs.io/en/latest/
+ */
+module ClickHouseDriver {
+  /** Gets a reference to the `clickhouse_driver` module. */
+  API::Node clickhouse_driver() { result = API::moduleImport("clickhouse_driver") }
+
+  /** Gets a reference to the `aioch` module. This module allows to make async db queries. */
+  API::Node aioch() { result = API::moduleImport("aioch") }
+
+  /**
+   * `clickhouse_driver` implements PEP249,
+   * providing ways to execute SQL statements against a database.
+   */
+  class ClickHouseDriverPEP249 extends PEP249ModuleApiNode {
+    ClickHouseDriverPEP249() { this = clickhouse_driver() }
+  }
+
+  module Client {
+    /** Gets a reference to a Client call. */
+    private DataFlow::Node client_ref() {
+      result = clickhouse_driver().getMember("Client").getASubclass*().getAUse()
+      or
+      result = aioch().getMember("Client").getASubclass*().getAUse()
+    }
```

Comment:
This part surprised me. Do you have an example of someone who provides their own Client subclass in a real project? (or is this just something that you thought _might_ handle and you wanted to capture?)

Reply:
I am sure i have seen few projects where there are custom subclasses of clickhouse-driver's Client. I did not save links to these projects. And actually dont remember how popular these projects are. I can spend some time on weekends to find them. I will put links here. And we will resolve this moment further.

What do you think? :)

Reply:
With a fast search, i have found this one https://github.com/Altinity/clickhouse-mysql-data-reader/blob/3b1b7088751b05e5bbf45890c5949b58208c2343/clickhouse_mysql/dbclient/chclient.py#L10
There are many stars, so here came the idea to support that kind of subclasses.

Reply:
Interesting, thanks for providing those examples :+1: having seen those, I think that it makes good sense to have support for that usage pattern :+1: